fdbserver/include/fdbserver/IEncryptionKeyProvider.actor.h

```diff
@@ -0,0 +1,281 @@
+/*
+ * IEncryptionKeyProvider.actor.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(NO_INTELLISENSE) && !defined(FDBSERVER_IENCRYPTIONKEYPROVIDER_ACTOR_G_H)
+#define FDBSERVER_IENCRYPTIONKEYPROVIDER_ACTOR_G_H
+#include "fdbserver/IEncryptionKeyProvider.actor.g.h"
+#elif !defined(FDBSERVER_IENCRYPTIONKEYPROVIDER_ACTOR_H)
+#define FDBSERVER_IENCRYPTIONKEYPROVIDER_ACTOR_H
+
+#include "fdbclient/GetEncryptCipherKeys.actor.h"
+#include "fdbclient/Tenant.h"
+#include "fdbserver/EncryptionOpsUtils.h"
+#include "fdbserver/ServerDBInfo.h"
+#include "flow/Arena.h"
+
+#define XXH_INLINE_ALL
+#include "flow/xxhash.h"
+
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+typedef uint64_t XOREncryptionKeyID;
+
+// EncryptionKeyRef is somewhat multi-variant, it will contain members representing the union
+// of all fields relevant to any implemented encryption scheme.  They are generally of
+// the form
+//   Page Fields - fields which come from or are stored in the Page
+//   Secret Fields - fields which are only known by the Key Provider
+// but it is up to each encoding and provider which fields are which and which ones are used
+struct EncryptionKeyRef {
+
+	EncryptionKeyRef(){};
+	EncryptionKeyRef(Arena& arena, const EncryptionKeyRef& toCopy)
+	  : cipherKeys(toCopy.cipherKeys), secret(arena, toCopy.secret), id(toCopy.id) {}
+	int expectedSize() const { return secret.size(); }
+
+	// Fields for AESEncryptionV1
+	TextAndHeaderCipherKeys cipherKeys;
+	Optional<BlobCipherEncryptHeader> cipherHeader;
+	// Fields for XOREncryption_TestOnly
+	StringRef secret;
+	Optional<XOREncryptionKeyID> id;
+};
+typedef Standalone<EncryptionKeyRef> EncryptionKey;
+
+// Interface used by pager to get encryption keys reading pages from disk
+// and by the BTree to get encryption keys to use for new pages
+class IEncryptionKeyProvider : public ReferenceCounted<IEncryptionKeyProvider> {
```

Comment:
a thought, if class is `templatized` then `BlobCipherKey` can be plugged directly as the return type avoiding need for `EncryptionKey` semantics. If `encryption` is enabled, then the class can be instantiated with `BlobCipherKey` specialization, otherwise, XOREncryption key semantics (if one exists). 

Thoughts?

Reply:
Templatize the IEncryptionKeyProvider will result in templatize much parts of Redwood, which is not a small change. I'm wanting to do a simpler clean up to interface making the XOR encryption use BlobCipherKey as well. But I will do it later. What do you think?

Reply:
I like the idea of BlobCipher to support XOR based encryption, however, following are my concerns:
1. If we go that route, then better to enhance BlobCipher to supply the encryption mode, today the only support mode is AES_256, we can add XOR as a supported mode and may be avoid code duplication in Redwood to maintain its own semantics of encryption?
2. BlobCipher supporting XOR also enables other modules to choose non-crypto secure (less CPU expensive) mechanism if needed (can’t think of a production usecase, but for us it would be a good extension to have (long term and not PuPr feature for sure)

Thoughts?

Reply:
Again, the discussion _need not_ to be part of this PR. We can do it offline as well. 
Thanks